run.ts

```diff
@@ -0,0 +1,135 @@
+import * as esbuild from "./deps/esbuild.ts"
+import { denoPlugin } from "./deps/esbuild_deno_loader.ts"
+import PQueue from "./deps/pqueue.ts"
+import { Browser } from "./deps/puppeteer.ts"
+import { deferred } from "./deps/std/async.ts"
+import { Buffer, readLines } from "./deps/std/io.ts"
+import * as path from "./deps/std/path.ts"
+import { readerFromStreamReader, writeAll } from "./deps/std/streams.ts"
+
+export interface RunOptions {
+  paths: (readonly [dir: string, fileName: string])[]
+  concurrency: number
+  runner: (dir: string, fileName: string) => Promise<void>
+  signal: AbortSignal
+}
+
+export async function run({ paths, concurrency, runner, signal }: RunOptions) {
+  const runQueue = new PQueue({ concurrency })
+  runQueue.addAll(paths.map(([dir, fileName]) => () => runner(dir, fileName)))
+
+  signal.addEventListener("abort", () => runQueue.clear())
+
+  await runQueue.onIdle()
+}
+
+export interface RunWithBrowserOptions {
+  createBrowser: () => Promise<Browser>
+  importMapUrl?: URL
+  results: (readonly [fileName: string, exitCode: number])[]
+}
+
+export async function runWithBrowser(
+  { createBrowser, importMapUrl, results }: RunWithBrowserOptions,
+) {
+  const browser = await createBrowser()
+  const consoleJs = await fetch(import.meta.resolve("./console.js")).then((r) => r.text())
+
+  return (async (dir: string, fileName: string) => {
+    const result = await executionWrapper(fileName, async (outputBuffer) => {
+      const page = await browser.newPage()
+      const buildResult = await esbuild.build({
+        plugins: [
+          denoPlugin({
+            importMapURL: importMapUrl,
+          }) as any,
+        ],
+        entryPoints: [path.join(dir, fileName)],
+        bundle: true,
+        write: false,
+        format: "esm",
+      })
+
+      await page.exposeFunction("__trun_injected_log", (...[msg]: [string, number]) => {
+        outputBuffer.writeSync(new TextEncoder().encode(msg))
+      })
+
+      const exitCode = deferred<number>()
+      await page.exposeFunction("exit", (args: string) => {
+        exitCode.resolve(Number(args))
+      })
+
+      const code = wrapCode(`${consoleJs}\n${buildResult.outputFiles[0]?.text!}`)
+      await page.addScriptTag({ content: code, type: "module" })
```

Comment:
```suggestion
      await page.addScriptTag({ content: consoleJs, type: "module" })
      const code = wrapCode(buildResult.outputFiles[0]?.text!)
      await page.addScriptTag({ content: code, type: "module" })
```

Reply:
Not sure if this will work but will give it a try

Reply:
Yup seems to be working. B/c its global i guess lol.